pkg/alertmanager/api.go

```diff
@@ -184,3 +185,42 @@ func validateUserConfig(logger log.Logger, cfg alertspb.AlertConfigDesc) error {
 
 	return nil
 }
+
+func (am *MultitenantAlertmanager) ListUserConfig(w http.ResponseWriter, r *http.Request) {
+	logger := util_log.WithContext(r.Context(), am.logger)
+	userIDs, err := am.store.ListAllUsers(r.Context())
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to list users of alertmanager")
+		http.Error(w, fmt.Sprintf("%s: %s", errListAllUser, err.Error()), http.StatusInternalServerError)
+		return
+	}
+	userConfigMap := make(map[string]*UserConfig, len(userIDs))
+	for _, userID := range userIDs {
+		cfg, err := am.store.GetAlertConfig(r.Context(), userID)
+		if err != nil {
+			if err == alertspb.ErrNotFound {
+				http.Error(w, err.Error(), http.StatusNotFound)
+			} else {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return
+		}
+		userConfigMap[userID] = &UserConfig{
+			TemplateFiles:      alertspb.ParseTemplates(cfg),
+			AlertmanagerConfig: cfg.RawConfig,
```

Comment:
Here I'd suggest writing partial output to response, and not keeping it in memory.

Reply:
partial output？
I don't quite understand this, can you elaborate more?

Reply:
Do you meaning output user configurations in batches, such as outputting 10 user configurations at a time?

Reply:
What I meant is marshalling and writing individual values to the response, which are then sent to client immediately, instead of building the entire `userConfigMap` in memory first. It would need to be done in such a way that what client see is single valid YAML document. I'm not sure how feasible that is.

In other words, instead of storing `userConfigMap[userID] = &UserConfig{...}`, we could marshal `&UserConfig{...}` into YAML and write it to response, possibly with some newlines. It's just an idea, and could be explored.

What worries me a bit is that we have some clusters with 1K+ tenants in Alertmanager (and it could easily be 10K+ in a year). Building response in memory for all those tenants could cause OOM.

Reply:
Get it, I will try.

Reply:
I have update to use partial output. It seems ok but only except that every output will add a '---'.

Reply:
update from yaml.NewEncoder to yaml.Marshal and everything seems good.

I think I have achieved: marshal single &UserConfig{...} into YAML and write it to response.